20190425-keras-premade-models.md

```diff
@@ -0,0 +1,380 @@
+| **Updated**   | 2019-04-25                                           |
+
+## Objective
+
+This document proposes several pre-made Keras models that would allow users to:
+* build basic machine learning models easily
+* compose them with other keras Layers
+* replace [Canned Estimators](https://www.tensorflow.org/guide/premade_estimators) in TF 2.0.
+
+## Motivation
+
+Scikit-learn is the de facto framework for creating basic machine learning models. From tensorflow perspective,
+it is beneficial to provide users several important basic models to possibly support the following features:
+* distributed mini-batch training
+* accelerated training (GPU & TPU)
+* enable AutoML to select different pre-made models
+* enable users to freely compose pre-made models with other Keras layers and models.
+
+Currently most pre-made models from tensorflow have been exposed through [CannedEstimator](https://www.tensorflow.org/guide/premade_estimators). However estimator
+* relies on continuous graph rebuilding and checkpoints reloading which slows down training
+* relies on global collections and not TF 2.0 friendly.
+* makes many advanced features such as meta/transfer learning difficult
+* enforces user to create input functions when not necessary
+
+Given the status, we propose the following pre-made models as a start:
+1. Linear models including Linear Regression and Logistic Regression
+2. DNN models including DNN Regression and DNN Classification
+3. Wide-Deep models including Wide-Deep Regression and Wide-Deep Classification
+4. BoostedTree models including BoostedTree Regression and BoostedTree Classification
+
+Users can use [feature_columns](https://www.tensorflow.org/guide/feature_columns) with pre-made models for feature engineering. We use **pre-made** and **canned** interchangeably in this doc.
+
+Several side notes: This document should serve as a foundation for other pre-made models that are not included in this proposal, such as Kmeans, random forest, collaborate filtering, which are yet to come; Detailed design regarding implementation details for BoostedTree, such as distributed training, saving & restoring, acceleration, supported feature columns should live in separate docs. 
+
+## Design Proposal
+
+### Challenges
+Many canned models, including BoostedTrees, KMeans (as well as WideDeep and many others more in the future) are highly complicated and do not follow the simple foward & backprop workflow, an assumption heavily relied on by Keras. Building such models, while not compromising the basic composability of Keras (i.e., compose layers on top of layers), is the main challenge we need to resolve in this document. Another challenge is that these models can have multiple training phases (such as collaborative filtering models). We propose the following two approaches to address them:
+
+### Proposal 1: Customized training function & composability
```

Comment:
What about other phases like evaluation and prediction? What's the difference between this and `tf.estimator`'s `model_fn`?

Reply:
evaluation and prediction are specified in Mode in Estimator world. In Keras model will by default create train_function, evaluate_function and predict_function and use the corresponding one during fit/eval/predict.